### PR TITLE
Minor PyTorch improvements

### PR DESCRIPTION
**Pull request by Anilith**
Hi,

First of all, thank you for publishing this wonderful library.
As the cluster I'm currently using only supports PyTorch I implemented some basic save/load/checkpoint functionalities and fixed the test set type in PyTorch (```RuntimeError: expected scalar type Double but found Float``` and ```RuntimeError: expected scalar type Float but found Double```).
If certain code sections need updating, I'll be open to fix them.

Kind regards,
Olivier

### CODE REVIEW

**Comment by lululxvi** [comment]
Awesome! Could you format your code by [black](https://github.com/psf/black)?

File: deepxde/geometry/geometry_nd.py [hunk 1]
```diff
@@ -68,7 +68,7 @@ def uniform_points(self, n, boundary=True):
                         1:
                     ]
                 )
-        x = np.array(list(itertools.product(*xi)))
+        x = np.array(list(itertools.product(*xi)), dtype=config.real(np))
```
Comment (lululxvi) [inline deepxde/geometry/geometry_nd.py:71]: How about enforcing the dtype in `np.linspace` as
```
if boundary:
    xi.append(np.linspace(self.xmin[i], self.xmax[i], num=ni, dtype=config.real(np)))
else:
    xi.append(np.linspace(self.xmin[i], self.xmax[i], num=ni + 1, endpoint=False, dtype=config.real(np))[1:])
```
So less memory is used.

**Comment by lululxvi** [comment]
By the way, your code has two parts: float/double fix, and save/load. You can make two separate pull requests, if you want.

File: deepxde/model.py [hunk 2]
```diff
@@ -594,32 +594,51 @@ def save(self, save_path, protocol="tf.train.Saver", verbose=0):
                 "tf.train.Saver" protocol supports ``restore()``.
         """
         # TODO: backend tensorflow, pytorch
-        if backend_name != "tensorflow.compat.v1":
+        if backend_name == "tensorflow.compat.v1": 
+            if verbose > 0:
+                print(
```
Comment (lululxvi) [inline deepxde/model.py:599]: The print info should be outside.

File: deepxde/model.py [hunk 3]
```diff
@@ -594,32 +594,51 @@ def save(self, save_path, protocol="tf.train.Saver", verbose=0):
...
+            if protocol == "tf.train.Saver":
+                self.saver.save(self.sess, save_path, global_step=self.train_state.epoch)
+            elif protocol == "pickle":
+                with open("{}-{}.pkl".format(save_path, self.train_state.epoch), "wb") as f:
+                    pickle.dump(self.state_dict(), f)
+
+        elif backend_name == "pytorch":
+                #torch.save does not make a directory, so directory has to be created before training
+                torch.save({
+                'epoch':  self.train_state.epoch,
```
Comment (lululxvi) [inline deepxde/model.py:613]: To be consistent with other backends:
- No need to save `epoch`.
- No need to save `loss`

File: deepxde/model.py [hunk 4]
```diff
@@ -594,32 +594,51 @@ def save(self, save_path, protocol="tf.train.Saver", verbose=0):
...
+                'model_state_dict': self.net.state_dict(),
+                'optimizer_state_dict': self.opt.state_dict(),
+                'loss':  self.train_state.loss_train,
+                }, save_path + "{}.pt".format(self.train_state.epoch))
```
Comment (lululxvi) [inline deepxde/model.py:617]: As discussed in https://pytorch.org/tutorials/beginner/saving_loading_models.html#saving-loading-a-general-checkpoint-for-inference-and-or-resuming-training, A common PyTorch convention is to save these checkpoints using the .tar file extension.

File: deepxde/model.py [hunk 5]
```diff
@@ -594,32 +594,51 @@ def save(self, save_path, protocol="tf.train.Saver", verbose=0):
...
+        if backend_name == "tensorflow.compat.v1":
+
+            if verbose > 0:
+                print("Restoring model from {} ...\n".format(save_path))
+            self.saver.restore(self.sess, save_path)
+        elif backend_name == "pytorch":
+            checkpoint = torch.load(save_path)
+            self.train_state.epoch = checkpoint["epoch"]
+            self.net.load_state_dict(checkpoint["model_state_dict"])
+            self.opt = checkpoint["optimizer_state_dict"]
```
Comment (lululxvi) [inline deepxde/model.py:636]: `self.opt.load_state_dict(checkpoint['optimizer_state_dict'])`?

File: deepxde/model.py [hunk 6]
```diff
@@ -594,32 +594,51 @@ def save(self, save_path, protocol="tf.train.Saver", verbose=0):
...
 
     def restore(self, save_path, verbose=0):
+        print("loading")
         """Restore all variables from a disk file."""
         # TODO: backend tensorflow, pytorch
-        if backend_name != "tensorflow.compat.v1":
+        if backend_name == "tensorflow.compat.v1":
+
+            if verbose > 0:
+                print("Restoring model from {} ...\n".format(save_path))
```
Comment (lululxvi) [inline deepxde/model.py:630]: The print info should be used for all backends

File: deepxde/model.py [hunk 7]
```diff
@@ -594,32 +594,51 @@ def save(self, save_path, protocol="tf.train.Saver", verbose=0):
...
-                )
-            )
-        if protocol == "tf.train.Saver":
-            self.saver.save(self.sess, save_path, global_step=self.train_state.epoch)
-        elif protocol == "pickle":
-            with open("{}-{}.pkl".format(save_path, self.train_state.epoch), "wb") as f:
-                pickle.dump(self.state_dict(), f)
 
     def restore(self, save_path, verbose=0):
+        print("loading")
```
Comment (lululxvi) [inline deepxde/model.py:624]: No need to use this one.

**Comment by Anilith** [comment]
Implemented your code suggestion for geometry_nd.py as it solves the problem, reformatted using black. I'm fixing the issues you mentioned in the code and opening a new pull request for saving and loading PyTorch models.

**Comment by lululxvi** [comment]
This pull request is split into two separate ones. For the PyTorch save/load, continue at https://github.com/lululxvi/deepxde/pull/437

**Pull request merged**